src/lib/containers/entity/annotations/AdditionalPropertiesSchemaField.tsx

```diff
@@ -158,20 +162,49 @@ export function AdditionalPropertiesSchemaField<T>(
     // The item may not be an array when we get it, and we need to convert it right away because the order of items is not stable, and seems to depend on if the item is an array or not.
     // Otherwise, the order of the properties will change when the user modifies the data. We may be able to fix this by modifying react-jsonschema-form to stabilize the item order.
 
-    // TODO: This doesn't work without a delay.
+    // FIXME: This doesn't work without a delay.
     setTimeout(() => {
       onChange(list)
-    }, 100)
+    }, 50)
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [])
 
+  /**
+   * This effect is invoked whenever the user attempts to change the data type of a custom annotation.
+   */
   useEffect(() => {
-    // When the selected type changes, switch to the appropriate widget for accepting input
-    setWidget(getWidgetFromPropertyType(propertyType))
+    function onNextPropertyTypeUpdate() {
+      const dataIsEmpty =
+        list.length === 0 || list.every(item => item == null || item == '')
+      const coercedList = transformDataFromPropertyType(list, nextPropertyType)
+      // if the data is empty or identical after conversion, then just update the property type
+      if (dataIsEmpty || nextPropertyType !== propertyType) {
+        if (isEqual(list, coercedList)) {
+          setPropertyType(nextPropertyType)
+        }
+      }
+    }
 
-    // Coerce the data to match the new type
-    setList(transformDataFromPropertyType(list, propertyType))
+    onNextPropertyTypeUpdate()
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [nextPropertyType])
 
+  /**
+   * This effect is invoked whenever the propertyType changes.
+   */
+  useEffect(() => {
+    function coerceDataAndUpdateWidget() {
+      const coercedList = transformDataFromPropertyType(list, nextPropertyType)
+
+      // Data conversion is non-destructive or has been confirmed by the user
+      setPropertyType(nextPropertyType)
```

Comment:
Not sure if this setPropertyType is necessary, I think the previous useEffect() or Confirm action takes care of this.  I think this useEffect could just reference propertyType, right?